### DISCUSSION TITLE
Setting field to aggregated value in update

### DISCUSSION THREAD

**Discussion by cj-olsen** [sqlalchemy/sqlalchemy#11413] [Usage Questions] [answered]
Hi there,

I have a Flask website where I'm trying to keep track of how many posts are in each category. The solution I'm trying to implement is having a `post_count` column in the category table that increments every time a new post is added to the category. I'd also like to have a function to recalculate the post counts in case they get out of sync.

Is it possible to set `post_count` to an aggregated value using an update statement? I searched online and read through the documentation for UPDATE and [UPDATE..FROM](https://docs.sqlalchemy.org/en/20/tutorial/data_update.html#update-from), but I wasn't able to find quite what I was looking for. The code below is what my project looks like right now.

```
from my_app.extensions.database import db
import sqlalchemy as sa
import sqlalchemy.orm as orm

class Category(db.Model):
    id: orm.Mapped[int] = orm.mapped_column(primary_key=True)
    name: orm.Mapped[str] = orm.mapped_column(sa.String(64), index=True, unique=True)
    post_count: orm.Mapped[int] = orm.mapped_column(default=0)
    posts: orm.WriteOnlyMapped['Post'] = orm.relationship(back_populates='category')

class Post(db.Model):
    id: orm.Mapped[int] = orm.mapped_column(primary_key=True)
    text: orm.Mapped[str] = orm.mapped_column(sa.Text())
    category_id: orm.Mapped[int] = orm.mapped_column(sa.ForeignKey(Category.id))
    category: orm.Mapped[Category] = orm.relationship(back_populates='posts')

def refresh_category_counts():
   stmt = (sa
       .update(Category, sa.func.count(Post.id).label('post_count'))
       .values(post_count='post_count'))
   
   db.session.execute(stmt)
   db.session.commit()
```

I am using SQLite 3 and SQLAlchemy 2.0.27.

Thanks in advance!

**Comment by CaselIT** [comment, marked as answer]
Hi,

It seems like you are looking for a subquery, see for example https://docs.sqlalchemy.org/en/20/tutorial/data_update.html#correlated-updates.
What you have I don't think it will work since you are not telling to filter the posts by category.

Also you could have a column property on category that returns you the count dynamically, instead of having to remember to run the update every time you add a post

**Reply by cj-olsen** [reply]
This is perfect, thank you! Here is what my function ended up looking like.

```
def refresh_post_counts():

    subquery = (sa
        .select(Category, sa.func.count(Post.id).label('post_count'))
        .join(Category.posts)
        .group_by(Category.id)
        .subquery())

    update_stmt = (sa
        .update(Category)
        .values(post_count=subquery.c.post_count)
        .where(Category.id == subquery.c.id))

    db.session.execute(update_stmt)
    db.session.commit()
```

I am using the count pretty frequently to e.g. calculate how many pages of posts are in a given category, so I was reluctant to get it dynamically on every page load. (Although if the performance impact is minimal maybe column properties would be more convenient?)

Thanks again for your help!
